src/Application/Actions/Donations/Update.php

```diff
@@ -258,10 +272,35 @@ private function addData(Donation $donation, HttpModels\Donation $donationData,
             }
 
             if ($donationData->autoConfirmFromCashBalance) {
-                $this->stripeClient->paymentIntents->confirm($donation->getTransactionId());
+                try {
+                    $this->stripeClient->paymentIntents->confirm($donation->getTransactionId());
+                } catch (InvalidRequestException $exception) {
+                    // Currently a typical Update call which auto-confirms is being made for just
+                    // that purpose. So our safest options are to return a 500 and roll back any
+                    // database changes.
+                    $this->entityManager->rollback();
+
+                    // To help analyse it quicker we handle the specific auto-confirm API failure we've
+                    // seen before with a distinct message, but both options give the client an HTTP 500,
+                    // as we expect neither with our updated guard conditions.
+                    if (str_contains($exception->getMessage(), 'missing a payment method')) {
+                        $this->logger->error(sprintf(
+                            'Stripe Payment Intent for donation ID %s was missing a payment method, so we could not confirm it',
+                            $donation->getUuid(),
+                        ));
+                        $error = new ActionError(ActionError::SERVER_ERROR, 'Could not confirm Stripe Payment Intent');
+
+                        return $this->respond($response, new ActionPayload(500, null, $error));
+                    }
+
+                    throw $exception;
+                }
             }
         }
 
+        // We hold off finishing the transaction for now, because the current logic
+        // involves rolling back our whole Update in *some* Stripe error scenarios.
+        // We don't have enough info to detect these without making the Stripe API call.
```

Comment:
I'm not sure I follow - if the transaction isn't finished here where does it get committed? I don't see any opportunity for a stripe error to happen after this line.

Reply:
Yep, not sure. I had this in my local uncommited work from before this week, and missed that it no longer makes sense with the current logic. Removed.